PXB-3220 tolerate file deletion between discovery and file open

aybek opened the pull request:
This PR contains fixes to
- PXB-3220 tolerate file deletion between discovery and file open
- PXB-3227 Dropped table is still in the backup
- PXB-3229 Can't create/write to file 't2.ibd.del' (OS errno 17 - File exists)
- PXB-3231 When handling renames: t1.ibd.ren file already exists

satya-bodapati reviewed (commented):
Looks good Aibek. Lets add some testcases with debug_sync.
with the debug sync point
1. drop a table between discovery and open
2. rename a table between discovery and open
3. ALTER TABLE (this isn't an issue, but to be safe)

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -52,5 +54,13 @@ class ddl_tracker_t {
+
+  /** Note that a table has been deleted between disovery and file open
+  @param[in]  file_path  missing table name with path. */
+  void add_missing_table(std::string path);
```
const std::string&

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.h:
```diff
@@ -52,5 +54,13 @@ class ddl_tracker_t {
+
+  /** Check if table is in missing list
+  @param[in]  name  tablespace name */
+  bool is_missing_table(std::string name);
```
const std::string&

satya-bodapati commented on storage/innobase/xtrabackup/src/ddl_tracker.cc:
```diff
@@ -226,8 +241,11 @@ void ddl_tracker_t::handle_ddl_operations() {
-    /* renamed new table. update new table entry to renamed table name */
-    if (new_tables.find(table.first) != new_tables.end()) {
+
+    /* renamed new table. update new table entry to renamed table name
+        || old table is missing add renamed table as a new table */
+    if (new_tables.find(table.first) != new_tables.end() ||
+        is_missing_table(table.second.first)) {
```
```suggestion
    /* renamed new table. update new table entry to renamed table name
    or if table is missing and renamed, add the renamed table to the new_table list. */
    if (new_tables.find(table.first) != new_tables.end() ||
        is_missing_table(table.second.first)) {
```

satya-bodapati replied:
Please elaborate the comment with an example

satya-bodapati commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -1610,8 +1610,8 @@ class Fil_system {
   [[nodiscard]] dberr_t iterate_spaces(Fil_space_iterator::Function &f);
-
 #endif /* XTRABACKUP */
+
```
I hope these new lines are not added intentionally. Please check

aybek committed:
PXB-3248 Multiple files found for the same tablespace ID

On prepare phase when handling ddl files some of the data files were not loaded to cache because of the first page validation
therefore were left without applying ddls on them.
To tolerate this issue we should open and load data files to cache without validation,to do so we are
using fil_tablespace_open_for_recovery() function instead of fil_open_for_xtrabackup().
